server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -176,26 +176,64 @@ public final void addDynamic(String name, String prefix, Mapper mapper, Document
-            } else {
-                // We strip off the first object path of the mapper name, load or create
-                // the relevant object mapper, and then recurse down into it, passing the remainder
-                // of the mapper name. So for a mapper 'foo.bar.baz', we locate 'foo' and then
-                // call addDynamic on it with the name 'bar.baz', and next call addDynamic on 'bar' with the name 'baz'.
-                int firstDotIndex = name.indexOf('.');
-                String immediateChild = name.substring(0, firstDotIndex);
-                String immediateChildFullName = prefix == null ? immediateChild : prefix + "." + immediateChild;
-                Builder parentBuilder = findObjectBuilder(immediateChildFullName, context);
-                if (parentBuilder != null) {
-                    parentBuilder.addDynamic(name.substring(firstDotIndex + 1), immediateChildFullName, mapper, context);
-                    add(parentBuilder);
-                } else if (subobjects.isPresent() && subobjects.get() == Subobjects.AUTO) {
-                    // No matching parent object was found, the mapper is added as a leaf - similar to subobjects false.
-                    add(name, mapper);
-                } else {
-                    // Expected to find a matching parent object but got null.
-                    throw new IllegalStateException("Missing intermediate object " + immediateChildFullName);
+                return;
+            }
+
+            if (subobjects.isPresent() && subobjects.get() == Subobjects.AUTO) {
+                // Check for parent objects. Due to auto-flattening, names with dots are allowed so we need to check for all possible
+                // object names. For instance, for mapper 'foo.bar.baz.bad', we have the following options:
+                // -> object 'foo' found => call addDynamic on 'bar.baz.bad'
+                // ---> object 'bar' found => call addDynamic on 'baz.bad'
+                // -----> object 'baz' found => add field 'bad' to it
+                // -----> no match found => add field 'baz.bad' to 'bar'
+                // ---> object 'bar.baz' found => add field 'bad' to it
+                // ---> no match found => add field 'bar.baz.bad' to 'foo'
+                // -> object 'foo.bar' found => call addDynamic on 'baz.bad'
+                // ---> object 'baz' found => add field 'bad' to it
+                // ---> no match found=> add field 'baz.bad' to 'foo.bar'
+                // -> object 'foo.bar.baz' found => add field 'bad' to it
+                // -> no match found => add field 'foo.bar.baz.bad' to parent
+                String fullPathToMapper = name.substring(0, name.lastIndexOf(mapper.leafName()));
+                String[] fullPathTokens = fullPathToMapper.split("\\.");
+                StringBuilder candidateObject = new StringBuilder();
+                String candidateObjectPrefix = prefix == null ? "" : prefix + ".";
```

Comment:
This is the most complex part of the pr.  I think this is covered the unit tests added in this test, but would be nice to see a one to one test for each of the above scenarios in the comment?

Reply:
Hm these tests would be quite verbose, with no extra testing coverage. I'd rather skip them, unless you feel strongly about it.

Reply:
For my education, why would these tests not provide extra coverage? Are all these scenarios already covered elsewhere?

Reply:
They're mostly covered in tests in `ObjectMapperTests` and `DynamicTemplatesTests`. Still, I think it's a good idea to refactor the logic (it's duplicated in 3 places at least) and properly test with randomization too.

Good thing is that we get very descent coverage from `StandardVersusLogsIndexModeRandomDataChallengeRestIT`, wrt randomized testing.